File: iocinfo.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const yaml = require('yaml');
/*
http://<beamline>-dbwr.<epik8namespace>/dbwr/view.jsp?display=https://<beamline>-docs..<epik8namespace>/control/<ioc>/<bobfile>
*/

/**
 * Convert URLs in text to markdown links
 */
function makeUrlsClickable(text) {
    const urlPattern = /(https?:\/\/[^\s]+)/g;
    return text.replace(urlPattern, '[$1]($1)');
}
/**
 * Extract metadata from start.log file
 */
function extractIocMetadata(startLogPath) {
    const metadata = {};
    if (fs.existsSync(startLogPath)) {
        const content = fs.readFileSync(startLogPath, 'utf8');
        const lines = content.split('\n');
        for (const line of lines) {
            if (line.includes(':')) {
                const [key, ...valueParts] = line.split(':');
                metadata[key.trim()] = valueParts.join(':').trim();
            }
        }
    }
    return metadata;
}

/**
 * Find a file recursively in a directory
 */
function findFile(dir, filename) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const fullPath = path.join(dir, file);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
            const found = findFile(fullPath, filename);
            if (found) return found;
        } else if (file === filename) {
            return fullPath;
        }
    }
    return null;
}

/**
 * Load the list of IOCs from values.yaml
 */
function loadIocList(valuesFile) {
    const content = fs.readFileSync(valuesFile, 'utf8');
    const values = yaml.parse(content);
    return {
        iocs: values.epicsConfiguration?.iocs || [],
        beamline: values.beamline || '',
        epik8namespace: values.epik8namespace || ''
    };
}

/**
 * Load services from values.yaml that have ingress enabled
 */
function loadServicesWithIngress(valuesFile) {
    const content = fs.readFileSync(valuesFile, 'utf8');
    const values = yaml.parse(content);
    
    const services = [];
    const beamline = values.beamline || '';
    const epik8namespace = values.epik8namespace || '';
    
    const epicsServices = values.epicsConfiguration?.services || {};
    for (const [serviceName, serviceConfig] of Object.entries(epicsServices)) {
        // Check if service has ingress enabled OR has a loadbalancer
        if (serviceConfig.enable_ingress || serviceConfig.ingress?.enabled || serviceConfig.loadbalancer) {
            // Determine the URL based on loadbalancer or standard URL format
            let url;
            if (serviceConfig.loadbalancer) {
                url = `http://${serviceConfig.loadbalancer}`;
            } else {
                const beamline = values.beamline || '';
                const epik8namespace = values.epik8namespace || '';
                url = `http://${beamline}-${serviceName}.${epik8namespace}`;
            }
            
            // Add path if specified
            if (serviceConfig.path) {
                url += serviceConfig.path;
            }
            
            services.push({
                name: serviceName,
                config: serviceConfig,
                url: url
            });
        }
    }
    
    return services;
}

/**
 * Generate IOC documentation pages
 */
function main(iocinfoDir, controlDir, valuesFile = null, servicesDir = 'content/services', opiDir = null) {
    // Load IOC list from values.yaml if provided
    let allowedIocs = null;
    let iocConfigs = {};
    let services = [];
    let beamline = '';
    let epik8namespace = '';
    if (valuesFile) {
        const data = loadIocList(valuesFile);
        const iocsData = data.iocs;
        beamline = data.beamline;
        epik8namespace = data.epik8namespace;
        allowedIocs = iocsData.map(ioc => ioc.name);
        iocConfigs = iocsData.reduce((acc, ioc) => {
            acc[ioc.name] = ioc;
            return acc;
        }, {});
        services = loadServicesWithIngress(valuesFile);
        console.log(`Filtering to IOCs listed in ${valuesFile}: ${allowedIocs}`);
        console.log(`Found ${services.length} services with ingress enabled`);
    }
    
    // Get all IOC directories
    const iocDirs = fs.readdirSync(iocinfoDir)
        .filter(d => fs.statSync(path.join(iocinfoDir, d)).isDirectory());
    
    // Filter IOCs if list is provided
    let filteredIocDirs = iocDirs;
    if (allowedIocs) {
        filteredIocDirs = iocDirs.filter(ioc => allowedIocs.includes(ioc));
        console.log(`Processing ${filteredIocDirs.length} IOCs: ${filteredIocDirs}`);
    }

    for (const iocName of filteredIocDirs) {
        const iocPath = path.join(iocinfoDir, iocName);
        const startLog = path.join(iocPath, 'start.log');
        
        // Define iocDestPath if valuesFile is provided
        let iocDestPath = null;
        if (valuesFile) {
            iocDestPath = path.join(controlDir, iocName);
        }
        
        // Extract metadata
        const metadata = extractIocMetadata(startLog);
        const devgroup = metadata['IOC Device Group'] || 'other';
        const iocAsset = metadata['IOC Asset'] || '';
        
        // Get IOC description if available
        let iocDesc = '';
        if (iocConfigs[iocName] && iocConfigs[iocName].desc) {
            iocDesc = iocConfigs[iocName].desc;
        }
        
        // Read start.log
        let startLogContent = '';
        if (fs.existsSync(startLog)) {
            startLogContent = makeUrlsClickable(fs.readFileSync(startLog, 'utf8'));
        }
        
        // Find YAML file
        const yamlFiles = fs.readdirSync(iocPath).filter(f => f.endsWith('.yaml'));
        let yamlContent = '';
        
        if (yamlFiles.length > 0) {
            const yamlPath = path.join(iocPath, yamlFiles[0]);
            yamlContent = fs.readFileSync(yamlPath, 'utf8');
        }
        
        // Read st.cmd if present
        let stcmdContent = '';
        const stcmdPath = path.join(iocPath, 'st.cmd');
        if (fs.existsSync(stcmdPath)) {
            stcmdContent = fs.readFileSync(stcmdPath, 'utf8');
        }
        
        // Read pvlist.txt
        let pvlistContent = '';
        let pvCount = 0;
        const pvlistPath = path.join(iocPath, 'pvlist.txt');
        if (fs.existsSync(pvlistPath)) {
            pvlistContent = fs.readFileSync(pvlistPath, 'utf8');
            pvCount = pvlistContent.trim().split('\n').filter(line => line.trim()).length;
        }
        
        // Check for .bob files
        let bobLink = '';
        const bobFiles = fs.readdirSync(iocPath).filter(f => f.endsWith('.bob'));
        if (bobFiles.length > 0 && beamline && epik8namespace) {
            const bobfile = bobFiles[0]; // Take the first one
            const url = `http://${beamline}-dbwr.${epik8namespace}/dbwr/view.jsp?display=https://${beamline}-docs.${epik8namespace}/control/${iocName}/${bobfile}`;
            bobLink = url;
        }
        
        // Check for OPI files from config
        const phoebusLinks = [];
        if (opiDir && iocConfigs[iocName]) {
            console.log(`Checking OPI for ${iocName}: opiDir=${opiDir}, opi=${iocConfigs[iocName]?.opi}`);
            if (iocConfigs[iocName].devices && iocConfigs[iocName].devices.length > 0) {
                // Per device
                for (const device of iocConfigs[iocName].devices) {
                    const deviceOpi = device.opi || iocConfigs[iocName].opi;
                    if (deviceOpi) {
                        let opiPath = path.join(opiDir, deviceOpi);
                        console.log(`Searching for OPI file for device ${device.name}: ${opiPath}`);
                        if (!fs.existsSync(opiPath)) {
                            // Try recursive search
                            const filename = path.basename(deviceOpi);
                            opiPath = findFile(opiDir, filename);
                            if (opiPath) {
                                console.log(`Found OPI file recursively: ${opiPath}`);
                            } else {
                                console.log(`OPI file not found: ${opiPath}`);
                                continue;
                            }
                        }
                        if (opiPath && fs.existsSync(opiPath)) {
                            const opiFile = path.basename(opiPath);
                            const relativePath = path.relative(controlDir, opiPath).replace(/\\/g, '/');
                            
                            // Create Phoebus link with macros
                            if (beamline && epik8namespace) {
                                let url = `http://${beamline}-dbwr.${epik8namespace}/dbwr/view.jsp?display=https://${beamline}-docs.${epik8namespace}/control/${relativePath}`;
                                const macros = {};
                                const iocprefix = iocConfigs[iocName].iocprefix || '';
                                if (iocprefix) {
                                    macros.P = iocprefix;
                                }
                                macros.R = device.name;
                                if (valuesFile) {
                                    macros.CONFFILE = valuesFile;
                                }
                                const macrosJson = JSON.stringify(macros);
                                const encodedMacros = encodeURIComponent(macrosJson);
                                url += `&macros=${encodedMacros}`;
                                phoebusLinks.push({ name: device.name, url: url });
                                console.log(`Created Phoebus link for ${device.name}: ${url}`);
                            }
                            
                            // No need to copy, link directly
                        }
                    }
                }
            }
            if (iocConfigs[iocName].opi && (!iocConfigs[iocName].devices || iocConfigs[iocName].devices.length === 0)) {
                console.log(`Condition met for IOC ${iocName}`);
                // IOC level
                let opiPath = path.join(opiDir, iocConfigs[iocName].opi);
                console.log(`Searching for OPI file for IOC ${iocName}: ${opiPath}`);
                if (!fs.existsSync(opiPath)) {
                    // Try recursive search
                    const filename = path.basename(iocConfigs[iocName].opi);
                    opiPath = findFile(opiDir, filename);
                    if (opiPath) {
                        console.log(`Found OPI file recursively: ${opiPath}`);
                    } else {
                        console.log(`OPI file not found: ${opiPath}`);
                    }
                }
                if (opiPath && fs.existsSync(opiPath)) {
                    const opiFile = path.basename(opiPath);
                    const relativePath = path.relative(controlDir, opiPath).replace(/\\/g, '/');
                    
                    // Create Phoebus link with macros
                    if (beamline && epik8namespace) {
                        let url = `http://${beamline}-dbwr.${epik8namespace}/dbwr/view.jsp?display=https://${beamline}-docs.${epik8namespace}/control/${relativePath}`;
                        const macros = {};
                        const iocprefix = iocConfigs[iocName].iocprefix || '';
                        if (iocprefix) {
                            macros.P = iocprefix;
                        }
                        macros.R = iocName;
                        if (valuesFile) {
                            macros.CONFFILE = valuesFile;
                        }
                        const macrosJson = JSON.stringify(macros);
                        const encodedMacros = encodeURIComponent(macrosJson);
                        url += `&macros=${encodedMacros}`;
                        phoebusLinks.push({ name: iocName, url: url });
                        console.log(`Created Phoebus link for ${iocName}: ${url}`);
                    }
                    
                    // No need to copy, link directly
                }
            }
        }
        
        // Check for .bob files in opi subdirectory
        const opiSubdir = path.join(iocPath, 'opi');
        if (fs.existsSync(opiSubdir)) {
            const bobFiles = fs.readdirSync(opiSubdir).filter(f => f.endsWith('.bob'));
            for (const bobFile of bobFiles) {
                if (beamline && epik8namespace) {
                    let url = `http://${beamline}-dbwr.${epik8namespace}/dbwr/view.jsp?display=https://${beamline}-docs.${epik8namespace}/control/${iocName}/${bobFile}`;
                    const macros = {};
                    const iocprefix = iocConfigs[iocName].iocprefix || '';
                    if (iocprefix) {
                        macros.P = iocprefix;
                    }
                    macros.R = iocName;
                    if (valuesFile) {
                        macros.CONFFILE = valuesFile;
                    }
                    const macrosJson = JSON.stringify(macros);
                    const encodedMacros = encodeURIComponent(macrosJson);
                    url += `&macros=${encodedMacros}`;
                    phoebusLinks.push({ name: bobFile.replace('.bob', ''), url: url });
                    console.log(`Created Phoebus link for ${bobFile} in opi subdir: ${url}`);
                }
            }
        }
        
        // Check for .bob files in global opiDir that match the IOC name
        if (opiDir) {
            const globalBobFiles = fs.readdirSync(opiDir).filter(f => f.endsWith('.bob') && f.toLowerCase().startsWith(iocName.toLowerCase()));
            for (const bobFile of globalBobFiles) {
                if (beamline && epik8namespace) {
                    const bobPath = path.join(opiDir, bobFile);
                    const relativePath = path.relative(controlDir, bobPath).replace(/\\/g, '/');
                    let url = `http://${beamline}-dbwr.${epik8namespace}/dbwr/view.jsp?display=https://${beamline}-docs.${epik8namespace}/control/${relativePath}`;
                    const macros = {};
                    const iocprefix = iocConfigs[iocName].iocprefix || '';
                    if (iocprefix) {
                        macros.P = iocprefix;
                    }
                    macros.R = iocName;
                    if (valuesFile) {
                        macros.CONFFILE = valuesFile;
                    }
                    const macrosJson = JSON.stringify(macros);
                    const encodedMacros = encodeURIComponent(macrosJson);
                    url += `&macros=${encodedMacros}`;
                    phoebusLinks.push({ name: bobFile.replace('.bob', ''), url: url });
                    console.log(`Created Phoebus link for global ${bobFile}: ${url}`);
                }
            }
        }
        
        // Generate markdown file
        const mdPath = path.join(controlDir, `${iocName}.md`);
        
        // Build navigation links
        const navLinks = [];
        if (iocAsset) {
            navLinks.push(`- [IOC Asset Documentation](${iocAsset})`);
        }
        if (phoebusLinks.length > 0) {
            navLinks.push('- [Phoebus Display](#phoebus)');
        }
        navLinks.push('- [Start Log](#startlog)');
        if (yamlContent) {
            navLinks.push('- [Configuration (YAML)](#yaml)');
        }
        if (stcmdContent) {
            navLinks.push('- [EPICS st.cmd](#stcmd)');
        }
        if (pvlistContent) {
            navLinks.push(`- [Process Variables (${pvCount})](#pvlist)`);
        }
        
        let yamlSection = '';
        if (yamlContent) {
            yamlSection = `
## Configuration (YAML) {#yaml}

\`\`\`yaml
${yamlContent}
\`\`\`
`;
        }
        
        let stcmdSection = '';
        if (stcmdContent) {
            stcmdSection = `
## EPICS st.cmd {#stcmd}

\`\`\`bash
${stcmdContent}
\`\`\`
`;
        }
        
        let pvlistSection = '';
        if (pvlistContent) {
            pvlistSection = `
## Process Variables (${pvCount}) {#pvlist}

\`\`\`text
${pvlistContent}
\`\`\`
`;
        }
        
        let descSection = '';
        if (iocDesc) {
            descSection = `
## Description

${iocDesc}

`;
        }
        
        let phoebusSection = '';
        if (phoebusLinks.length > 0) {
            phoebusSection = `
## Phoebus Display {#phoebus}

${phoebusLinks.map(link => `- [${link.name}](${link.url})`).join('\n')}

`;
        }
        
        const mdContent = `---
title: "${iocName}"
linkTitle: "${iocName}"
weight: 10
devgroup: "${devgroup}"
type: docs
date: "${new Date().toISOString()}"
---

## Quick Navigation

${navLinks.join('\n')}

## Start Log {#startlog}

\`\`\`
${startLogContent}
\`\`\`
${descSection}${phoebusSection}${yamlSection}${stcmdSection}${pvlistSection}
`;
        
        // Ensure control directory exists
        if (!fs.existsSync(controlDir)) {
            fs.mkdirSync(controlDir, { recursive: true });
        }
        
        fs.writeFileSync(mdPath, mdContent);
        
        // Copy IOC directory to control directory only if valuesFile is provided (filtered IOCs)
        if (valuesFile) {
            // iocDestPath already defined above
            if (!fs.existsSync(iocDestPath)) {
                fs.mkdirSync(iocDestPath, { recursive: true });
            }
            // Copy files from iocPath to iocDestPath
            const allowedExtensions = ['.bob', '.yaml', '.txt', '.log', '.md','.html','.opi'];
            const files = fs.readdirSync(iocPath);
            for (const file of files) {
                const srcFile = path.join(iocPath, file);
                const destFile = path.join(iocDestPath, file);
                if (fs.statSync(srcFile).isFile() && allowedExtensions.includes(path.extname(srcFile))) {
                    fs.copyFileSync(srcFile, destFile);
                }
            }
            
            // Copy opi subdirectory if it exists
            const opiSrcPath = path.join(iocPath, 'opi');
            const opiDestPath = path.join(iocDestPath, 'opi');
            if (fs.existsSync(opiSrcPath)) {
                if (!fs.existsSync(opiDestPath)) {
                    fs.mkdirSync(opiDestPath, { recursive: true });
                }
                const opiFiles = fs.readdirSync(opiSrcPath).filter(f => allowedExtensions.includes(path.extname(f)));
                for (const file of opiFiles) {
                    fs.copyFileSync(path.join(opiSrcPath, file), path.join(opiDestPath, file));
                }
            }
            
            // Copy matched global .bob files - no longer needed, link directly
            // if (opiDir) {
            //     const globalBobFiles = fs.readdirSync(opiDir).filter(f => f.endsWith('.bob') && f.toLowerCase().startsWith(iocName.toLowerCase()));
            //     for (const bobFile of globalBobFiles) {
            //         const srcFile = path.join(opiDir, bobFile);
            //         const destFile = path.join(iocDestPath, bobFile);
            //         if (fs.existsSync(srcFile)) {
            //             fs.copyFileSync(srcFile, destFile);
            //         }
            //     }
            // }
        }
        
        const assetInfo = iocAsset ? ` | Asset: ${iocAsset.substring(0, 40)}...` : '';
        console.log(`Updated: ${mdPath} (PVs: ${pvCount})${assetInfo}`);
    }

    // Generate documentation for services with ingress
    for (const service of services) {
        const serviceName = service.name;
        const serviceUrl = service.url;
        const serviceConfig = service.config;
        
        // Get service description if available
        const serviceDesc = serviceConfig.desc || 'This is an EPICS service with ingress enabled.';
        
        // Determine if this is a loadbalancer-only service (no ingress)
        const hasIngress = serviceConfig.enable_ingress || serviceConfig.ingress?.enabled;
        const hasLoadbalancer = !!serviceConfig.loadbalancer;
        
        const mdPath = path.join(servicesDir, `${serviceName}.md`);
        
        // Format the connection info based on ingress availability
        let connectionInfo;
        if (hasLoadbalancer && !hasIngress) {
            // Load balancer only - show IP address
            connectionInfo = `**Connection IP:** ${serviceConfig.loadbalancer}`;
        } else {
            // Ingress enabled - show clickable URL
            connectionInfo = `## Service URL

[${serviceUrl}](${serviceUrl})`;
        }
        
        const mdContent = `---
title: "${serviceName}"
linkTitle: "${serviceName}"
weight: 10
type: docs
date: "${new Date().toISOString()}"
---

${connectionInfo}

## Description

${serviceDesc}
`;
        
        // Ensure services directory exists
        if (!fs.existsSync(servicesDir)) {
            fs.mkdirSync(servicesDir, { recursive: true });
        }
        
        fs.writeFileSync(mdPath, mdContent);
        
        console.log(`Updated: ${mdPath} (Service: ${serviceUrl})`);
    }

    console.log('Done!');
}

// Parse command line arguments
if (require.main === module) {
    const args = process.argv.slice(2);
    let iocinfoDir = 'static/iocinfo';
    let controlDir = 'content/control';
    let servicesDir = 'content/services';
    let valuesFile = null;
    let opiDir = null;
    
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--iocinfo-dir' && i + 1 < args.length) {
            iocinfoDir = args[i + 1];
            i++;
        } else if (args[i] === '--control-dir' && i + 1 < args.length) {
            controlDir = args[i + 1];
            i++;
        } else if (args[i] === '--services-dir' && i + 1 < args.length) {
            servicesDir = args[i + 1];
            i++;
        } else if (args[i] === '--values-file' && i + 1 < args.length) {
            valuesFile = args[i + 1];
            i++;
        } else if (args[i] === '--opi-dir' && i + 1 < args.length) {
            opiDir = args[i + 1];
            i++;
        } else if (args[i] === '--help' || args[i] === '-h') {
            console.log('Usage: node iocinfo.js [options]');
            console.log('');
            console.log('Options:');
            console.log('  --iocinfo-dir <path>   Directory containing IOC information (default: static/iocinfo)');
            console.log('  --control-dir <path>   Output directory for control documentation (default: content/control)');
            console.log('  --services-dir <path>  Output directory for services documentation (default: content/services)');
            console.log('  --values-file <path>   Path to values.yaml file to filter IOCs by epicsConfiguration.iocs list');
            console.log('  --opi-dir <path>       Directory containing OPI files referenced in IOC configurations');
            console.log('  --help, -h             Show this help message');
            process.exit(0);
        }
    }
    
    main(iocinfoDir, controlDir, valuesFile, servicesDir, opiDir);
}

module.exports = { main, extractIocMetadata, makeUrlsClickable };

File: iocinfo.py
import os
import re
import argparse
import yaml
from datetime import datetime

def extract_ioc_metadata(start_log_path):
    """Extract metadata from start.log file"""
    metadata = {}
    if os.path.exists(start_log_path):
        with open(start_log_path, 'r') as f:
            for line in f:
                if ':' in line:
                    key, value = line.split(':', 1)
                    metadata[key.strip()] = value.strip()
    return metadata

def make_urls_clickable(text):
    """Convert URLs in text to markdown links"""
    url_pattern = r'(https?://[^\s]+)'
    return re.sub(url_pattern, r'[\1](\1)', text)

def load_ioc_list(values_file):
    """Load the list of IOCs from values.yaml"""
    with open(values_file, 'r') as f:
        values = yaml.safe_load(f)
    return values.get('epicsConfiguration', {}).get('iocs', [])

def load_services_with_ingress(values_file):
    """Load services from values.yaml that have ingress enabled"""
    with open(values_file, 'r') as f:
        values = yaml.safe_load(f)
    
    services = []
    beamline = values.get('beamline', '')
    epik8namespace = values.get('epik8namespace', '')
    
    epics_services = values.get('epicsConfiguration', {}).get('services', {})
    for service_name, service_config in epics_services.items():
        # Check if service has ingress enabled OR has a loadbalancer
        if service_config.get('enable_ingress') or service_config.get('ingress', {}).get('enabled') or 'loadbalancer' in service_config:
            # Determine the URL based on loadbalancer or standard URL format
            if 'loadbalancer' in service_config:
                url = f"http://{service_config['loadbalancer']}"
            else:
                beamline = values.get('beamline', '')
                epik8namespace = values.get('epik8namespace', '')
                url = f"http://{beamline}-{service_name}.{epik8namespace}"
            
            # Add path if specified
            if 'path' in service_config:
                url += service_config['path']
            
            services.append({
                'name': service_name,
                'config': service_config,
                'url': url
            })
    
    return services

def main(iocinfo_dir, control_dir, values_file=None, services_dir=None):
    """Generate IOC documentation pages"""
    # Load IOC list from values.yaml if provided
    allowed_iocs = None
    ioc_configs = {}
    services = []
    if values_file:
        iocs_data = load_ioc_list(values_file)
        allowed_iocs = [ioc['name'] for ioc in iocs_data]
        ioc_configs = {ioc['name']: ioc for ioc in iocs_data}
        services = load_services_with_ingress(values_file)
        print(f"Filtering to IOCs listed in {values_file}: {allowed_iocs}")
        print(f"Found {len(services)} services with ingress enabled")
    
    # Get all IOC directories
    ioc_dirs = [d for d in os.listdir(iocinfo_dir) 
                if os.path.isdir(os.path.join(iocinfo_dir, d))]
    
    # Filter IOCs if list is provided
    if allowed_iocs:
        ioc_dirs = [ioc for ioc in ioc_dirs if ioc in allowed_iocs]
        print(f"Processing {len(ioc_dirs)} IOCs: {ioc_dirs}")

    for ioc_name in ioc_dirs:
        ioc_path = os.path.join(iocinfo_dir, ioc_name)
        start_log = os.path.join(ioc_path, "start.log")
        
        # Extract metadata
        metadata = extract_ioc_metadata(start_log)
        devgroup = metadata.get('IOC Device Group', 'other')
        ioc_asset = metadata.get('IOC Asset', '')
        
        # Read start.log
        start_log_content = ""
        if os.path.exists(start_log):
            with open(start_log, 'r') as f:
                start_log_content = make_urls_clickable(f.read())
        
        # Find YAML file
        yaml_files = [f for f in os.listdir(ioc_path) if f.endswith('.yaml')]
        yaml_content = ""
        
        if yaml_files:
            yaml_path = os.path.join(ioc_path, yaml_files[0])
            with open(yaml_path, 'r') as f:
                yaml_content = f.read()
        
        # Read st.cmd if present
        stcmd_content = ""
        stcmd_path = os.path.join(ioc_path, "st.cmd")
        if os.path.exists(stcmd_path):
            with open(stcmd_path, 'r') as f:
                stcmd_content = f.read()
        
        # Read pvlist.txt
        pvlist_content = ""
        pv_count = 0
        pvlist_path = os.path.join(ioc_path, "pvlist.txt")
        if os.path.exists(pvlist_path):
            with open(pvlist_path, 'r') as f:
                pvlist_content = f.read()
                pv_count = len([line for line in pvlist_content.strip().split('\n') if line.strip()])
        
        # Generate markdown file
        md_path = os.path.join(control_dir, f"{ioc_name}.md")
        
        # Get IOC description if available
        ioc_desc = ""
        if ioc_name in ioc_configs and 'desc' in ioc_configs[ioc_name]:
            ioc_desc = ioc_configs[ioc_name]['desc']
        
        # Build navigation links
        nav_links = []
        if ioc_asset:
            nav_links.append(f"- [IOC Asset Documentation]({ioc_asset})")
        nav_links.append("- [Start Log](#startlog)")
        if yaml_content:
            nav_links.append("- [Configuration (YAML)](#yaml)")
        if stcmd_content:
            nav_links.append("- [EPICS st.cmd](#stcmd)")
        if pvlist_content:
            nav_links.append(f"- [Process Variables ({pv_count})](#pvlist)")
        
        yaml_section = ""
        if yaml_content:
            yaml_section = f"""
## Configuration (YAML) {{#yaml}}

```yaml
{yaml_content}
```
"""
        
        stcmd_section = ""
        if stcmd_content:
            stcmd_section = f"""
## EPICS st.cmd {{#stcmd}}

```bash
{stcmd_content}
```
"""
        
        pvlist_section = ""
        if pvlist_content:
            pvlist_section = f"""
## Process Variables ({pv_count}) {{#pvlist}}

```text
{pvlist_content}
```
"""
        
        desc_section = ""
        if ioc_desc:
            desc_section = f"""
## Description

{ioc_desc}

"""
        
        md_content = f"""---
title: "{ioc_name}"
linkTitle: "{ioc_name}"
weight: 10
devgroup: "{devgroup}"
date: "{datetime.now().isoformat()}"
---

## Quick Navigation

{chr(10).join(nav_links)}

## Start Log {{#startlog}}

```
{start_log_content}
```
{desc_section}{yaml_section}{stcmd_section}{pvlist_section}
"""
        
        with open(md_path, 'w') as f:
            f.write(md_content)
        
        asset_info = f" | Asset: {ioc_asset[:40]}..." if ioc_asset else ""
        print(f"Updated: {md_path} (PVs: {pv_count}){asset_info}")

    # Generate documentation for services with ingress
    for service in services:
        service_name = service['name']
        service_url = service['url']
        service_config = service['config']
        
        # Get service description if available
        service_desc = service_config.get('desc', 'This is an EPICS service with ingress enabled.')
        
        # Determine if this is a loadbalancer-only service (no ingress)
        has_ingress = service_config.get('enable_ingress') or service_config.get('ingress', {}).get('enabled')
        has_loadbalancer = 'loadbalancer' in service_config
        
        md_path = os.path.join(services_dir, f"{service_name}.md")
        
        # Format the connection info based on ingress availability
        if has_loadbalancer and not has_ingress:
            # Load balancer only - show IP address
            connection_info = f"**Connection IP:** {service_config['loadbalancer']}"
        else:
            # Ingress enabled - show clickable URL
            connection_info = f"## Service URL\n\n[{service_url}]({service_url})"
        
        md_content = f"""---
title: "{service_name}"
linkTitle: "{service_name}"
weight: 10
type: docs
date: "{datetime.now().isoformat()}"
---

{connection_info}

## Description

{service_desc}
"""
        
        # Ensure services directory exists
        os.makedirs(services_dir, exist_ok=True)
        
        with open(md_path, 'w') as f:
            f.write(md_content)
        
        print(f"Updated: {md_path} (Service: {service_url})")

    print("Done!")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Generate IOC documentation pages from iocinfo directory')
    parser.add_argument('--iocinfo-dir', default='static/iocinfo',
                        help='Directory containing IOC information (default: static/iocinfo)')
    parser.add_argument('--control-dir', default='content/control',
                        help='Output directory for control documentation (default: content/control)')
    parser.add_argument('--services-dir', default='content/services',
                        help='Output directory for services documentation (default: content/services)')
    parser.add_argument('--values-file', 
                        help='Path to values.yaml file to filter IOCs by epicsConfiguration.iocs list')
    
    args = parser.parse_args()
    main(args.iocinfo_dir, args.control_dir, args.values_file, args.services_dir)
